### PR TITLE
[roles:srv-announcer] Add basic role structure and tasks to install executable

### PR DESCRIPTION
**Pull request by lucendio**
The role implements 3 main steps:

1. fetch and executable and bring it in place
2. confiugure systemd service
3. start and test the service

Ii is worth noting that for now the `srv-announcer` can only manage one SRV record, whereas it might be necessary to announce items on multiple SRV records (e.g. `_sft.` and `_sft-metrics.`). The role allows to configure this, which will result in multiple instances of the `srv-announcer` but with slightly different configuration. For more information, please refer to the comments next to the corresponding variable `announcer_records`.

### CODE REVIEW

File: roles/srv-announcer/templates/srv-announcer.service.j2
Comment (flokli) [inline roles/srv-announcer/templates/srv-announcer.service.j2:17]: We should never put secrets inside unit files. Use `EnvironmentFile`, and place secrets in the file referenced there.

File: roles/srv-announcer/vars/main.yml
Comment (flokli) [inline roles/srv-announcer/vars/main.yml:1]: I don't understand the constants/not constants distinction TBH.

Given `announcer_artifact_basename` contains the version, does this mean one can't ever update the version?

File: roles/srv-announcer/tasks/start.yml
Comment (flokli) [inline roles/srv-announcer/tasks/start.yml:14]: This whole `deprecated` thing seems to be a new concept, or do we use the same approach elsewhere already?

I'd personally just have stopped the announcer service (via `systemctl stop`) every time we do maintainance, some hours before decomissioning the server entirely - Do we really want to track intermediate server states during decomissioning in version control?

File: roles/srv-announcer/templates/srv-announcer.service.j2
Comment (flokli) [reply roles/srv-announcer/templates/srv-announcer.service.j2:17]: Also, https://github.com/systemd/systemd/pull/16568 might be relevant once merged (and it doesn't hurt to be aware of it before)

Comment (lucendio) [reply roles/srv-announcer/templates/srv-announcer.service.j2]: [Done](https://github.com/wireapp/ansible-sft/pull/8/commits/e38b884e14a237ab0fb11d67670196026dc044f4).

File: roles/srv-announcer/vars/main.yml
Comment (lucendio) [reply roles/srv-announcer/vars/main.yml:1]: The idea here is that while [`announcer_version`](https://github.com/wireapp/ansible-sft/pull/8/files#diff-00ee6cc36478d9ce7f01fb88808ddb9fR12) can (and should) be defined by the user of this role, `announcer_artifact_basename`, on the other hand, is defined by the role author or even the entity in charge of defining the [name of the artifact](https://github.com/wireapp/srv-announcer/blob/master/.github/workflows/release-and-upload.yml#L37).

Distinguishing based on the use cases led to putting certain vars either in `./defaults` or `./vars` to make the ones that should not be touched by the user de facto read-only.

Does this makes it a bit more clear?

File: roles/srv-announcer/tasks/start.yml
Comment (lucendio) [reply roles/srv-announcer/tasks/start.yml:16]: I'm not sure if we have the same use case in mind. Could you elaborate a bit, please? Also, what happens (or what would you like to happen) when running `systemctl stop`?

File: roles/srv-announcer/vars/main.yml
Comment (jschaul) [reply roles/srv-announcer/vars/main.yml:1]: Can we please move everything underneath vars/main.yml to defaults/main.yml?

We don't gain anything from hiding these variables and making them impossible to override. Clearly, for version v0.1.1 of the announcer, we want to override the shasum. And this is a very valid usecase as a `user` of the role, rather than as a role maintainer. Blocking your users of this role from using a newer artifact until you as the maintainer got the time to add a shasum and a new release doesn't make sense to me.

this is just one example. It's hard to know in advance where users may have a need to override a variable. Requiring a role update for any change is not a good practice. Let's please not make our lives more complicated than they need to be; and move these variables to defaults/main.yml for flexibility. If a role user 'mistakenly' overrides a variable that you don't want them to override, it's their problem, not yours. So why forbid this override?

File: roles/srv-announcer/vars/main.yml
Comment (jschaul) [inline roles/srv-announcer/vars/main.yml:3]: please prefix all variables with the same prefix, e.g. make this announcer_service_name. Or, for consistency, use `srv_announcer_` as a prefix everywhere.

File: roles/srv-announcer/meta/main.yml
Comment (jschaul) [inline roles/srv-announcer/meta/main.yml:1]: This role cannot be downloaded by ansible galaxy in its current form. Even with collections, the required folder paths don't match. Therefore, why do you include a galaxy file here?

File: roles/srv-announcer/defaults/main.yml
Comment (jschaul) [inline roles/srv-announcer/defaults/main.yml:1]: This comment is not news and can be removed. Anyone reading the ansible documentation knows that variables under defaults/main can be overriden.

File: meta/main.yml
Comment (jschaul) [inline meta/main.yml:1]: How is this file used? What ansible-galaxy command can or cannot make use of this file?

File: roles/srv-announcer/tasks/finalize.yml
Comment (flokli) [inline roles/srv-announcer/tasks/finalize.yml:7]: This has the same remarks as my [previous comment](https://github.com/wireapp/ansible-sft/pull/8/files#r471424361)

File: roles/srv-announcer/tasks/start.yml
Comment (flokli) [reply roles/srv-announcer/tasks/start.yml:16]: If we want to stop the announcer to stop getting new connections, just stop the service.

If it's too inconvenient to do it manually, provide a playbook just stopping this (which can be limited to certain hosts).

If we want to radically change naming of these services, in ansible world we're probably better off by entirely redeploying the machine, instead of dealing with error-prone multi-staged removal phases.

File: roles/srv-announcer/vars/main.yml
Comment (lucendio) [reply roles/srv-announcer/vars/main.yml:1]: Apparently I did a bad job in explaining/documenting the behaviour you are questioning.

There is no need of overriding a checksum. The user has [complete flexibility](https://github.com/wireapp/ansible-sft/blob/add-srv-announcer-role/roles/srv-announcer/defaults/main.yml#L6-L16) in determining the source of the artifact. 

Case 1: if the user provides a path to the artifact, it is assumed that she already verified its integrity before invoking the role. After all, when providing a path the artifact has already been downloaded

Case 2: the user only provides a version, therefore the role needs to be aware of the corresponding checksum, hence [this](https://github.com/wireapp/ansible-sft/blob/add-srv-announcer-role/roles/srv-announcer/vars/main.yml#L17-L19) ever growing map of version --> checksum

Case 3: the user wants to handover the download part to the role, therefore a URL and a checksum are needed. A slightly different scenario - version + checksum - could easily be done outside of the role by deriving the URL from interpolating the version into a URL template string. 

Overall, those cases should cover every scenario. Or would you have one in mind that isn't? 

If you argue against this approach, of using constants, you could question private variables of a class or public vs. private APIs as well - same pattern in that sense.

I can see your point regarding flexibility, although with some reasonable design, I think this won't be the case much often. I actually belief that having both makes it more clear to the reader/user to a) discover the parameters she is actually looking for and wants to provide, and the ones that are internal and would only be noise when looking for these kind of things.

To conclude, if I'm still not able to make a valid case here, I'd be happy to put everything together. Let me know.

File: roles/srv-announcer/tasks/start.yml
Comment (lucendio) [reply roles/srv-announcer/tasks/start.yml:16]: There are probably two pieces to it.

>  [...] just stop the service.

That is correct as far as I understand the tool. The "health check" needs to fail in order to make the announcer remove the entry. When the announcer itself is stopped, it wont remove the entry in the SRV record list it is responsible for - at least up until now.

The operational task to start/stop the service, that the announcer is responsible for, is [not yet implemented](https://github.com/wireapp/ansible-sft/blob/develop/roles/sft-server/tasks/start.yml#L5), but would probably be helpful due to the current announcer behaviour. What do you think? (you already proposed the playbook, I think it could also just be a flag).
Another way would be to extend the `srv-announcer` so that it would depublish the list item when being stopped (or expose a one-off sub-command that can be hooked into ` ExecStop=`).

Furthermore, the role currently assumes that multiple announcer instances (one for each SRV record <--> list item, e.g. FQDN) run on one machine. The reasoning was to let the announcer also maintain the the metrics SRV record. But, since we decided to maintain that one In Terraform, the overall approach might not longer be valid. Anyway, the [role API](https://github.com/wireapp/ansible-sft/blob/add-srv-announcer-role/roles/srv-announcer/defaults/main.yml#L40) for this could still be useful, no?

The `deprecated` flag is to facilitate add/removal `srv-announcer` instances on a machine over time. If neither multiple instances on one machine nor add/removal over time are valid use cases, the `loop`s for the unit file and service tasks could be removed. To facility the use case of managing multiple SRV record <--> FQDN pairs on one machine -if desired - the `srv-announcer` could be extended to take in more than just one pair.

File: roles/srv-announcer/vars/main.yml
Comment (lucendio) [reply roles/srv-announcer/vars/main.yml]: [Done](https://github.com/wireapp/ansible-sft/pull/8/commits/37a1bfb601755e53dcb37343f52b7956a9cd4543)

File: meta/main.yml
Comment (lucendio) [reply meta/main.yml:1]: The existence of this file allows to download the role(s) via `ansible-galaxy`, it just needs a [small workaround](https://github.com/zinfra/cailleach/blob/master/environments/exxon-dev/ansible/sft/ansible.cfg#L8). A shart explanation can be found in the last bullet of [this](https://github.com/zinfra/cailleach/commit/5701e7412d2fc07fcd3b5b1a65b0dace09aacf2c) commit message.

File: roles/srv-announcer/meta/main.yml
Comment (lucendio) [reply roles/srv-announcer/meta/main.yml:1]: While [my comment below](https://github.com/wireapp/ansible-sft/pull/8#discussion_r478476774) might give some context, you are indeed correct, right now this file has no technical impact, it just holds some, IMHO, useful meta information about the role itself. Would you like me to do anything about this file?

File: roles/srv-announcer/defaults/main.yml
Comment (lucendio) [reply roles/srv-announcer/defaults/main.yml:1]: In my opinion it's valuable to put a fact from the docs into the code that actually explains why this file exists and how its content should be understood. If you think, for example, of on-prem or just people using what we provide w/o specific domain knowledge.

Also, I'm a bit confused, because you signed of [#1](https://github.com/wireapp/ansible-sft/pull/1) which contained a [similar comment](https://github.com/wireapp/ansible-sft/pull/1/files#diff-ee5d19d9692ad3b7ab8374167a65ca48R1) w/o highlighting this.

Comment (jschaul) [reply roles/srv-announcer/defaults/main.yml:1]: If you think documentation for end users is needed, I think a good place would be on docs.wire.com, and/or inside the README. If you know enough ansible to know that defaults/main.yml contain overridable variables, and then open that file, you don't need to be told this anymore. So I feel this comment isn't very useful.

> Also, I'm a bit confused, because you signed of #1 which contained a similar comment w/o highlighting this.

I generally only comment on the most important issues on a given PR and let less important things pass. Also, I don't necessarily review every line of code or comment in detail. I noticed it here, so I commented on it. Having received an approval for 659 lines of code doesn't mean I fully agree 100% with each of those lines of code.

But it's a comment, so I won't continue a discussion here anymore, if you feel strongly to keep it, keep it.

File: meta/main.yml
Comment (jschaul) [reply meta/main.yml:1]: It would be great if "how to download this with ansible-galaxy (with a workaround)" was documented either in this file, or, even better, in the README of this github repository.

Knowing that this file exists for a workaround without knowing the workaround isn't very useful. Please link to (or write) publicly accessible documentation for this.

File: roles/srv-announcer/meta/main.yml
Comment (jschaul) [reply roles/srv-announcer/meta/main.yml:1]: Add a comment then, stating that this role cannot be used by ansible galaxy, but that you chose to use the ansible galaxy meta format to describe that this role runs on ubuntu 18... and needs ansible 2.7? It's rather odd.

Why not add this information to the README, and delete this file?

File: roles/srv-announcer/defaults/main.yml
Comment (jschaul) [inline roles/srv-announcer/defaults/main.yml]: Perhaps we can keep priority at 0 for consistency? Or is there a motivation for this value to be different to how we use priority in other places?

File: roles/srv-announcer/vars/main.yml
Comment (jschaul) [reply roles/srv-announcer/vars/main.yml:1]: > Case 2: the user only provides a version, therefore the role needs to be aware of the corresponding checksum, hence this ever growing map of version --> checksum

Well... this is inside vars, so not overridable. If a new release happens in in github for srv announcer, then a user of this role is unable to use option 2 with a newer version of srv announcer. This is a problematic restriction (it's a restriction, it has downsides) and I don't see the benefit of this "constants" approach here. Your comment doesn't explain *why* it's useful to disallow ansible role users from overriding something. What is the benefit? You say - if I get this right - in the scenario where a newer version of an artifact exists and the role isn't aware yet, if option 2 cannot be used, well then the user is free to use option 3 instead. That's nice, but why make things more complicated than they need to be? 

But what about if I want to always have the `latest` version available, whatever that might be?

Other examples:

* srv_announcer is hosted on gitlab going forward. This role has implicit assumptions that github is where srv announcer is found (it's hardcoded here).
* If I fork the srv announcer, I can currently not override the srv_announcer_download_base_url. I would need to *also* fork and maintain this ansible-sft project.
* If I fork the srv announcer, but decide to rename it, I cannot override the artifact basename without first forking this repo
* I want to support 32 bit. Using my forked srv announncer repo. But I cannot do this, I first need to fork this, and move srv_announcer_arch to defaults/main.yml, and then I can override it in group_vars for the 32-bit webserver group

But all this is beside the point. I was trying to come up with a reason to *not restrict users* by giving an example. The example is just that - an example. I cannot really foresee what someone might want to do in 2021 with this role. However, I have in the past been quite frustrated with ansible role developers who unnecessarily placed variables inside /vars for no good reason, and I had to fork the project, instead of being able to simply override one variable - which is already there, in variable form (not hardcoded in 3 places), but it's in an unoverridable location. So I would like us to not start a pattern in which it's hard to override things without forking the code.

> If you argue against this approach, of using constants, you could question private variables of a class or public vs. private APIs as well - same pattern in that sense.

Ansible is not really a programming language, it's a limited DSL with a range of conventions and helpers. It's not the same.

I found another person with a view on this that I find sensible, see for example https://codereviewvideos.com/course/ansible-tutorial/video/variable-precedence-where-to-put-your-role-vars

Is this another case of https://github.com/zinfra/backend-issues/issues/1617 ?

File: roles/srv-announcer/vars/main.yml
Comment (jschaul) [reply roles/srv-announcer/vars/main.yml]: Thanks!

File: roles/srv-announcer/vars/main.yml
Comment (jschaul) [reply roles/srv-announcer/vars/main.yml:1]: (I also opened a separate discussion for this here: https://github.com/zinfra/backend-issues/issues/1707 )

Comment (lucendio) [reply roles/srv-announcer/vars/main.yml:1]: For Case 2 - only provide a version - I cannot see a way to make it work w/o hard-coding and therefore releasing a new version of the role every time a new SFT version comes out, because of the checksum. Fetching the checksum (e.g. `get_url`) is pointless.

I think, most of your concerns in that regard can be addressed (or at least be minimized to an absolute minimum) by proper automation timing these things (e.g. artifact publishing and checksum adding + version bumping in the role). 

`latest`: If that is a desired use case - and I hope it's not, because it leads to unpredictable behaviour - this could easily be solved with *case 3* + us providing a URL where the version is replaced by *latest* always pointing to the latest release (similar to container images and the *latest* tag). Tho, it's far from being practical, because of the ever-changing checksum. I intention would be to prevent people from doing these things.

For me, it's about providing the user with a maximum amount of convenience while keeping it flexible, hence `srv_announcer_version`.

>> If you argue against this approach, of using constants, you could question private variables of a class or public vs. private APIs as well - same pattern in that sense.

> Ansible is not really a programming language, it's a limited DSL with a range of conventions and helpers. It's not the same.

There are differences, for sure. But, IMHO, there are also similarities and applicable concepts. Let's agree to disagree :)

> To conclude, if I'm still not able to make a valid case here, I'd be happy to put everything together. Let me know.

I'm not going into every example you brought up, but I see your point. Even though the offer still stands, I'd much rather like to discuss your concerns about specific variables defined in `vars/main.yml` - like `srv_announcer_download_base_url` - instead of just moving everything over.

File: roles/srv-announcer/defaults/main.yml
Comment (lucendio) [reply roles/srv-announcer/defaults/main.yml]: I did't look into other places. I just wanted to reflect the [defaults](https://github.com/wireapp/srv-announcer/blob/master/main.go#L30-L107).

File: meta/main.yml
Comment (lucendio) [reply meta/main.yml:1]: You are right. I tried to explain [this](https://github.com/wireapp/ansible-sft/pull/8/commits/22ad187aeffe49ec803e3016a28da7083747b9eb).

File: roles/srv-announcer/meta/main.yml
Comment (lucendio) [reply roles/srv-announcer/meta/main.yml:1]: The roles can indeed be downloaded via `ansible-galaxy`, as mentioned [here](https://github.com/wireapp/ansible-sft/pull/8#discussion_r480253787).

I don't feel that this odd. The role meta information seem to be the one (and machine-readable) location for such information. At least, it's where I would assume people look first.

> Why not add this information

Those information are already mentioned in the [README.md](https://github.com/wireapp/ansible-sft/blame/add-srv-announcer-role/README.md#L4-L12).

File: roles/srv-announcer/tasks/configure.yml
Comment (flokli) [inline roles/srv-announcer/tasks/configure.yml:25]:
```suggestion
  register: unit_srv_announcer
```

Unrelated, but I stumbled it over during review. Making this string a bit longer doesn't hurt, especially considering how many other characters we remove ;-)

**Review by flokli — APPROVED** [review]
A small nit for clarity, apart from that, LGTM!

File: roles/srv-announcer/tasks/configure.yml
Comment (lucendio) [reply roles/srv-announcer/tasks/configure.yml]: [Addressed](https://github.com/wireapp/ansible-sft/pull/8/commits/8ec9040f0dc18c921b5f2faa6170af6c55b4728b).

**Pull request merged**
